pkg/cmd/server/admin/create_mastercerts.go

```diff
@@ -180,65 +212,73 @@ func (o CreateMasterCertsOptions) Validate(args []string) error {
 func (o CreateMasterCertsOptions) CreateMasterCerts() error {
 	glog.V(4).Infof("Creating all certs with: %#v", o)
 
-	signerCertOptions := CreateSignerCertOptions{
-		CertFile:   DefaultCertFilename(o.CertDir, CAFilePrefix),
-		KeyFile:    DefaultKeyFilename(o.CertDir, CAFilePrefix),
-		SerialFile: DefaultSerialFilename(o.CertDir, CAFilePrefix),
-		ExpireDays: o.SignerExpireDays,
-		Name:       o.SignerName,
-		Overwrite:  o.Overwrite,
-		Output:     o.Output,
-	}
-	if err := signerCertOptions.Validate(nil); err != nil {
-		return err
-	}
-	if _, err := signerCertOptions.CreateSignerCert(); err != nil {
-		return err
-	}
-	// once we've minted the signer, don't overwrite it
-	getSignerCertOptions := SignerCertOptions{
-		CertFile:   DefaultCertFilename(o.CertDir, CAFilePrefix),
-		KeyFile:    DefaultKeyFilename(o.CertDir, CAFilePrefix),
-		SerialFile: DefaultSerialFilename(o.CertDir, CAFilePrefix),
+	if hasPhase(PhaseSign, o.Phases) {
+		signerCertOptions := CreateSignerCertOptions{
+			CertFile:   o.SignerCertOptions.CertFile,
+			KeyFile:    o.SignerCertOptions.KeyFile,
+			SerialFile: o.SignerCertOptions.SerialFile,
+			ExpireDays: o.SignerExpireDays,
+			Name:       o.SignerName,
+			Overwrite:  o.Overwrite,
+			Output:     o.Output,
+		}
+		if err := signerCertOptions.Validate(nil); err != nil {
+			return err
+		}
+		if _, err := signerCertOptions.CreateSignerCert(); err != nil {
+			return err
+		}
 	}
 
 	errs := parallel.Run(
-		func() error { return o.createCABundle(&getSignerCertOptions) },
-		func() error { return o.createServerCerts(&getSignerCertOptions) },
-		func() error { return o.createAPIClients(&getSignerCertOptions) },
-		func() error { return o.createEtcdClientCerts(&getSignerCertOptions) },
-		func() error { return o.createKubeletClientCerts(&getSignerCertOptions) },
-		func() error { return o.createProxyClientCerts(&getSignerCertOptions) },
+		func() error { return o.createCABundle(o.SignerCertOptions) },
+		func() error { return o.createServerCerts(o.SignerCertOptions) },
+		func() error { return o.createAPIClients(o.SignerCertOptions) },
+		func() error { return o.createEtcdClientCerts(o.SignerCertOptions) },
+		func() error { return o.createKubeletClientCerts(o.SignerCertOptions) },
+		func() error { return o.createProxyClientCerts(o.SignerCertOptions) },
 		func() error { return o.createServiceAccountKeys() },
-		func() error { return o.createServiceSigningCA(&getSignerCertOptions) },
+		func() error { return o.createServiceSigningCA() },
 	)
 	return utilerrors.NewAggregate(errs)
 }
 
 func (o CreateMasterCertsOptions) createAPIClients(getSignerCertOptions *SignerCertOptions) error {
+	caFiles := []string{}
+	// Include the signer cert if specified and exists
+	if len(getSignerCertOptions.CertFile) > 0 {
+		if _, err := os.Stat(getSignerCertOptions.CertFile); err == nil || !os.IsNotExist(err) {
```

Comment:
The `!os.IsNotExist` condition looks suspicious. Do we really need to include cert if the file non-readable for the current user, for example?